### DISCUSSION TITLE
GitHub actions for adding old versions to OpenUPM

### DISCUSSION THREAD

**Discussion by kezzyhko** [openupm/openupm#5201] [Show and tell]
There are some [guides from OpenUPM](https://openupm.com/docs/managing-upm-project.html) on maintaining a package. I had a usecase of maintaining a forked repo, which already had its own releases. The guides suggested either automatic release creation on push to main branch (not suitable for me because it's a fork with its own tags), or an action to sync fork with upstream without automatic releases (I don't want to create release manually especially because upstream repo already has them). So, I have created GitHub Actions that syncs fork with upstream, merges existing tags into new branch, and automatically creates releases on corresponding merge commits. This comes with the added benefit of releasing not only new versions to OpenUPM, but all of the previous ones too.

Here's how it works:
* You fork the repo
* You create new branch from `main`, say `upm`. Branch should be created from earliest commit possible, preferrably before first release.
* You add needed changes to `upm`
* You add workflows to `upm`
* Each 12 hours, sync action is triggered (first time can be triggered manually)
* This sync action merges upstream into your fork
* Then sync action checks for new tags
* For each new tag, sync action runs a release action
* Release action converts tag to a valid semver string
* Release action merges from the original tag into your `upm` branch
* Release action creates a release at new merge commit.
Tag of this release will be a correct semver string generated previously, prefixed with `upm-` to avoid collisions
It will contain both your changes and changes from upstream repo.
* If there are more tags from the upstream repo without a release, sync action will continue running and run release action once more

Also, I decided to have additional `upm-dev` branch, and merge all changes from it to `upm`. This is needed for the edge case of actions running while `upm` branch is in unsable state and creating broken release. You can also set `upm` to default branch. Basically, `upm` becomes your new main, and `main` becomes just a regular branch for getting updates from upstream.

Examples for reference: https://github.com/kezzyhko/mujoco-upm and https://github.com/kezzyhko/mj-unity-extensions-upm

**Comment by favoyang** [comment]
@kezzyhko thanks for sharing this approach, and demonstrating it with an example repo.
